baremetal/src/include/ebbrt/GeneralPurposeAllocator.h

```diff
@@ -15,6 +15,33 @@
 #include <ebbrt/VMemAllocator.h>
 
 namespace ebbrt {
+
+// page fault handler for mapping in physical pages
+// to virtual pages on all cores
+class LargeRegionFaultHandler : public ebbrt::VMemAllocator::PageFaultHandler {
+  std::map<uintptr_t, ebbrt::Pfn> tmap;
+  size_t ps = 2 * 1024 * 1024;  // large page size
+
+ public:
+  void SetMap(std::map<uintptr_t, ebbrt::Pfn> m) { tmap = std::move(m); }
+
+  // given faulted address, calculates virtual page frame and finds
+  // corresponding physical page frame in tmap table, then calls
+  // MapMemory to do actual mapping
+  void HandleFault(ebbrt::idt::ExceptionFrame* ef,
+                   uintptr_t faulted_address) override {
+    auto vpage = ebbrt::Pfn::LDown(faulted_address);
+    auto it = tmap.find(vpage.ToLAddr());
+    if (likely(it != tmap.end())) {
+      ebbrt::vmem::MapMemoryLarge(it->first, it->second, ps);
+    } else {
+      kabort("%s: abort in large region fault handler, cannot find 0x%lx in "
+             "map table.\n",
+             __PRETTY_FUNCTION__, vpage.ToLAddr());
+    }
+  }
+};
```

Comment:
But I think you can store the beginning vaddr, the length, and then just compute the index into a vector rather than performing a lookup into a map.